### DISCUSSION TITLE
VBAF -- Educational AI and RL Framework in Pure PowerShell 5.1 (DQN, PPO, A3C from scratch)

### DISCUSSION THREAD

**Discussion by JupyterPS** [PowerShell/PowerShell#27639] [Show and tell]
Hi PowerShell Team and community,

I wanted to share something I have been building for the past year -- VBAF, 
the only complete AI and reinforcement learning framework written entirely 
in PowerShell 5.1.

No Python. No dependencies. Runs on any Windows PC with no setup beyond:

    Install-Module VBAF -Scope CurrentUser

What it implements from scratch:
- Neural networks with backpropagation (Rumelhart 1986)
- Q-Learning with Q-table (Watkins 1989)
- Deep Q-Networks with experience replay and target network (Mnih 2015)
- Proximal Policy Optimization -- same family as ChatGPT training (Schulman 2017)
- Asynchronous Advantage Actor-Critic (Mnih 2016)
- Multi-agent market simulation with emergent behaviours
- 14 enterprise automation pillars built on DQN

Every algorithm includes comments explaining the mathematics in plain English 
with references to the original research papers. The code is the curriculum.

I noticed your 2026 investments post mentions AI integration with PowerShell 
as a priority. VBAF demonstrates that serious AI/RL work is already possible 
in pure PowerShell 5.1 today -- and could serve as a reference or educational 
resource for the community.

GitHub: https://github.com/JupyterPS/VBAF
PSGallery: https://www.powershellgallery.com/packages/VBAF

Happy to answer any questions about the implementation.

Henning -- Roskilde, Denmark

**Comment by JosephColvin** [comment]
I have to say, after reviewing just the NN I feel like I'm back in HS learning about them. Nice base and really good explanation and education (citing the sources are a welcome sight). This shows your effort and I really have never seen this done before with NN and AI. I would love to see you rename the variable "error."

**Reply by JupyterPS** [reply]
Thank you Joseph -- great catch! $error is a PowerShell automatic variable and should never be used as a user variable name. Renamed to $sampleLoss throughout the training loop and pushed. Exactly the kind of feedback that makes VBAF better.
